api/envoy/extensions/filters/http/dynamic_modules/v3/dynamic_modules.proto

```diff
@@ -0,0 +1,45 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.dynamic_modules.v3;
+
+import "envoy/extensions/dynamic_modules/v3/dynamic_modules.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.dynamic_modules.v3";
+option java_outer_classname = "DynamicModulesProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/dynamic_modules/v3;dynamic_modulesv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: HTTP filter for dynamic modules]
+
+// Configuration of the HTTP filter for dynamic modules. This filter allows loading shared object files
+// that can be loaded via dlopen by the HTTP filter.
+//
+// A module can be loaded by multiple HTTP filters, hence the program can be structured in a way that
+// the module is loaded only once and shared across multiple filters providing multiple functionalities.
+//
+// Currently, the implementation is work in progress and not usable.
+message DynamicModuleFilter {
+  // Specifies the shared-object level configuration.
+  envoy.extensions.dynamic_modules.v3.DynamicModuleConfig dynamic_module_config = 1;
```

Comment:
Does it really make sense for the control plane to have to know the actual path to the dynamic module, or will that be something local controlled by the workload or deployment?  I'm wondering if it would be valuable to have a layer of indirection here, where the bootstrap file defines some dynamic module instances, giving each one a name, and the control plane sends the instance name of the module to use.  This way, the API between the control plane and the workload is defined in terms of these abstract instance names rather than in terms of specific paths.

Reply:
I think this is a good question. In fact, for the case of Wasm, even though it has the same DataSource config like this PR, the reality of Istio is that the control plane always passes a local path and the sidecar agent is the one downloading the binary from remote source and place it at the location specified by the control plane because the Envoy's filter chain set up routine cannot be made asynchronous, hence the filter chain would end up in a dangerous filter state (e.g. downloading failed but the filter chain initialized).

For the case of dynamic modules, now I think at least `DataSource` configuration is overkill since unlike Wasm case, `dlopen` can only refer a local file and cannot use the in-memory bytes. So at the end of the day, we would end up using local file paths in reality just like the Wasm case, e.g. Envoy would need to write `inline_bytes` into a temporary files etc (which I think is messy tbh).

Regarding your idea of using the bootstrap file stuff, if I understand correctly, we would lose the ability to load/distribute module files at runtime without restarting Envoy vs if we can specify the local file path here, at least we can achieve that with the help of a side car agent etc.

From implementation point of view, there's almost no difference between these choices,, so I think the question is here is whether or not we want to tie the Envoy deployment with loadable dynamic modules by having a dynamic modules specific config in the Envoy bootstrap config, or have the ability to load/discover modules at runtime. wdyt? @mattklein123 @marc-barry 

cc @arkodg @zirain @zhaohuabing

Reply:
I _guess_ in practice, users wouldn't distribute native binaries at runtime as it's tied to the machine arch/os Envoy is running on, and therefore they would deploy the modules together with Envoy. So, personally either choice works for me.

Reply:
so basically what I am thinking right now is to stop using `DataSource` (the reason I described above) and have a dedicated `oneof` so that we can later add another option like the bootstrap config thingy discussed here like
```diff
diff --git a/api/envoy/extensions/dynamic_modules/v3/dynamic_modules.proto b/api/envoy/extensions/dynamic_modules/v3/dynamic_modules.proto
index 204c482b02..5e612e2a7c 100644
--- a/api/envoy/extensions/dynamic_modules/v3/dynamic_modules.proto
+++ b/api/envoy/extensions/dynamic_modules/v3/dynamic_modules.proto
@@ -36,8 +36,15 @@ option (xds.annotations.v3.file_status).work_in_progress = true;
 // Currently, the implementation is work in progress and not usable.
 // [#next-free-field: 8]
 message DynamicModuleConfig {
-  // The location of the object file. Currently, only the local path is supported.
-  config.core.v3.DataSource object_file = 7;
+  // The location of the object file. Intentionally this does not use config.core.v3.DataSource, as the
+  // object file must be loaded via dlopen, which requires a physical file path, not a in-memory buffer.
+  oneof location {
+    // The object file's local path. It can be a relative path or an absolute path.
+    // E.g. /path/to/module.so or module.so
+    string local_path = 2;
+
+    // TODO: maybe a "name" pointing to a file configured in the bootstrap.
+  }
 
   // Set true to prevent the module from being unloaded with dlclose.
   // This is useful for modules that have global state that should not be unloaded.
```

Reply:
Note that the [xDS API style guidelines](https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md) discourage use of `oneof`.

Reply:
Note that since my main concern here was avoiding the need for the control plane to know local paths that may be deployment- or workload-specific, switching from `DataSource` to `string local_path` doesn't really help with this.

If we agree that we want a layer of indirection here to isolate the control plane from local paths, then maybe we should just go ahead and do that from the start?  It doesn't seem that hard to do, unless I'm missing something.

Reply:
> If we agree that we want a layer of indirection here to isolate the control plane from local paths,

This is what I am not sure about, though I understand the motivation. For example, two main envoy control planes Envoy Gateway and Istio both deploy its sidecar agents to Envoy deployment, and they basically manage some local volumes along with the xds configurations with local paths. So, I am not sure how much "hiding local paths" here help the implementations - at the end of the day "control planes" (including agents) needs to be aware of local paths.

Reply:
having said that, I think I am fine with: an env var to specify the search directory (just like LD_LIBRARY_PATH on linux, and defaults the current working directory maybe?) and only a name being specified in xds API here. And then Envoy searches `lib${name}.so` in the directory. This matches the default Linux behavior but I want to hear how others feel about it.

Reply:
yeah actually I probably like it better than the local paths specified in xds.

Reply:
ok I addressed the feedback here: https://github.com/envoyproxy/envoy/pull/36448/commits/86315d784a727333c96a467b3d4c55f9d078f45c